pylmnn/impostor_sampling.py

```diff
@@ -0,0 +1,164 @@
+"""Tools to assist in sampling impostors for PyLMNN. """
+from abc import ABC, abstractmethod
+import numpy as np
+
+
+class Sampler(ABC):
+    @abstractmethod
+    def extend(self, latest_values):
+        pass
+
+    @abstractmethod
+    def sample(self):
+        pass
+
+
+class UniformSampler(Sampler):
+    """Post-hoc uniform sample of fixed size from a stream values.
+
+    Parameters
+    ----------
+    sample_size : int
+        Number of elements to take as the sample.
+
+    random_state : numpy.RandomState
+        A pseudo random number generator object used for sampling.
+    """
+
+    def __init__(self, sample_size, random_state):
+        self.sample_size = sample_size
+        self.random_state = random_state
+
+        self._values = []
+
+    def extend(self, latest_values):
+        """Adds elements to be sampled from.
+
+        Parameters
+        ----------
+        latest_values : array-like, shape (n_samples,)
+            The latest part of the stream to sample from.
+
+        """
+        self._values.extend(latest_values)
+
+    def sample(self):
+        """Gets the current sample.
+
+        Returns
+        -------
+        sample : array, shape (sample_size,)
+            The current sample from the stream
+
+        """
+        n = len(self._values)
+        k = self.sample_size
+        if n > k:
+            ind = self.random_state.choice(n, k, replace=False)
+            return np.asarray(self._values, dtype=object)[ind]
```

Comment:
What is the overhead here, when values contains both distances and indices? Maybe there is a more efficient way to sample from multiple arrays? We could also use the knowledge that there are always either 1 or 2 arrays used.

Reply:
I agree, this is not ideal, and you're right the root cause is needing to support sampling from parallel arrays of different types (just indicies or indicies and distances). I saw this as a bigger problem and didn't address any of it in favor of brevity.